benchmarks/bench_micro.c

```diff
@@ -0,0 +1,333 @@
+    ACTION(pct_put,         OPTION_TYPE_UINT, 10,    "% of puts")\
+    ACTION(pct_rem,         OPTION_TYPE_UINT, 10,     "% of removes")
+
+#define O(b, opt) option_uint(&(b->options.benchmark.opt))
+
+typedef size_t benchmark_key_t;
+
+struct benchmark_entry {
+    char *key;
+    benchmark_key_t key_size;
+
+    char *value;
+    size_t value_size;
+};
+
+struct benchmark_specific {
+    BENCHMARK_OPTION(OPTION_DECLARE);
+};
+
+struct benchmark_options {
+    struct benchmark_specific benchmark;
+    cuckoo_options_st cuckoo;
+};
+
+struct benchmark {
+    FILE *config;
+    struct benchmark_entry *entries;
+    struct benchmark_options options;
+};
+
+static int
+benchmark_create(struct benchmark *b, const char *config)
+{
+    b->entries = NULL;
+
+    struct benchmark_specific opts1 = { BENCHMARK_OPTION(OPTION_INIT) };
+    cuckoo_options_st opts2 = { CUCKOO_OPTION(OPTION_INIT) };
+    b->options.benchmark = opts1;
+    b->options.cuckoo = opts2;
+
+    size_t nopts = OPTION_CARDINALITY(struct benchmark_options);
+    option_load_default((struct option *)&b->options, nopts);
+
+    if (config != NULL) {
+        b->config = fopen(config, "r");
+        if (b->config == NULL) {
+            log_crit("failed to open the config file");
+            return -1;
+        }
+        option_load_file(b->config, (struct option *)&b->options, nopts);
+        fclose(b->config);
+    }
+
+    if (O(b, entry_min_size) <= sizeof(benchmark_key_t)) {
+        log_crit("entry_min_size must larger than %lu",
+            sizeof(benchmark_key_t));
+
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+benchmark_destroy(struct benchmark *b)
+{
+}
+
+static struct benchmark_entry
+benchmark_entry_create(benchmark_key_t key, size_t size)
+{
+    struct benchmark_entry e;
+    e.key_size = sizeof(key);
+    e.value_size = size - sizeof(key);
+    e.key = cc_alloc(e.key_size);
+    ASSERT(e.key != NULL);
+    e.value = cc_alloc(e.value_size);
+    ASSERT(e.value != NULL);
+
+    memcpy(e.key, &key, e.key_size);
+    e.key[e.key_size - 1] = 0;
+
+    memset(e.value, 'a', e.value_size);
+    e.value[e.value_size - 1] = 0;
+
+    return e;
+}
+
+static void
+benchmark_entry_destroy(struct benchmark_entry *e)
+{
+    cc_free(e->key);
+    cc_free(e->value);
+}
+
+static void
+benchmark_entries_populate(struct benchmark *b)
+{
+    size_t nentries = O(b, nentries);
+    b->entries = cc_alloc(sizeof(struct benchmark_entry) * nentries);
+    ASSERT(b->entries != NULL);
+
+    for (size_t i = 1; i <= nentries; ++i) {
+        size_t size = RRAND(O(b, entry_min_size), O(b, entry_max_size));
+        b->entries[i - 1] = benchmark_entry_create(i, size);
+    }
+}
+
+static void
+benchmark_entries_delete(struct benchmark *b)
+{
+    for (size_t i = 0; i < O(b, nentries); ++i)
+        benchmark_entry_destroy(&b->entries[i]);
+    cc_free(b->entries);
+}
+
+static int
+benchmark_cuckoo_init(struct benchmark *b)
+{
+#define CUCKOO_ITEM_OVERHEAD 32
```

Comment:
i believe there is already an ITEM_OVERHEAD macro defined in cuckoo/item.h. or is this different?